### PR TITLE
Added logging logic for mood recipe changes, part 2

### LINKED ISSUE #1411: a mood die which is the target of a trip attack and changes to the same size doesn't say so
Example:
defender_no_forced_size_message

It should be:

```
c1 performed Trip attack using [t(1):1] against [o(Z=20)?:12]; Attacker t(1) rerolled 1 => 1; Defender o(Z=20)? remained the same size, rerolled 12 => 10, was not captured
```

### LINKED ISSUE #1663: Doppelganger vs. Mad/Mood
When a Doppleganger die captures a Mad or Mood die it rerolls the copied die. That reroll should trigger a potential resize of the die but currently does not. For example:

> blackshadowshade performed Power attack using [D(10):7] against [(X=12)?:7]; Defender (X=12)? 
> was captured; Attacker D(10) changed size from 10 to 12 sides, recipe changed from D(10) to 
> (X=12)?, rerolled 7 => 12

### PR DESCRIPTION
**Pull request by blackshadowshade**
This is part two of #1675. Fixes #1411. Fixes #1663.

### CODE REVIEW

File: src/engine/BMGameAction.php [hunk 1]
```diff
@@ -687,24 +687,68 @@ protected function message_size_change(array $preInfo, array $postInfo) {
     /**
      * Describes a die recipe change
      *
+     * There are six possible scenarios here:
+     * - no recipe change
+     * - a non-mood recipe change
+     * - a mood recipe no change
+     * - a mood recipe change
+     * - a non-mood recipe change with a mood recipe no change
+     * - a non-mood recipe change with a mood recipe change
```
Comment (cgolubi1) [inline src/engine/BMGameAction.php:696]: Nit: why not put examples for each of these, e.g.:

```
* - a mood recipe no change: `(X=6)? => (X=6)?`
```

Reading this initially, i'm not sure what the last two items mean --- i'm sure it's obvious and my brain's just not online yet, but an example would make it obvious regardless.

**Comment by cgolubi1** [comment]
You have conflicts to resolve on this (presumably with the subdie logic).  I'm going to kick off a replay site anyway, to get some initial info, and then i'll rebuild it once you take care of that.

**Comment by cgolubi1** [comment]
Oh, right, i remember how this works now: i can't trivially kick off a replay site when there are merge conflicts with master.   So i'll let you fix that first, and then i'll start replay testing.

**Comment by blackshadowshade** [comment]
This has been rebased to resolve conflicts in the responder tests, and I've added examples to the comment.

**Comment by cgolubi1** [comment]
Getting the flag changes out of this made a big difference in terms of testability, but there's still two things going on here:
* Changes in the text which is displayed when various things happen
* Logic changes in what happens during an attack

An example of the latter --- right now replay testing is failing out on this because it expects extra die rolls:

```
        $this->verify_api_submitTurn(
            array(3, 4),
            'responder003 performed Power attack using [D(12):11] against [s(R=4,R=4)?:5]; Defender s(R=4,R=4)? was captured; Attacker D(12) changed size from 12 to 8 sides, recipe changed from D(12) to s(R=4,R=4)?, rerolled 11 => 7. ',
            $retval, array(array(0, 2), array(1, 3)),
            $gameId, 2, 'Power', 0, 1, '', array());
```

Anyway, i'm going to continue moving forward with trying to test, and i'm making progress, but one question i have is: from your point of view, if we split again, starting by trying to merge everything *except* the one-liner changes to `src/engine/BMSkillDoppelganger.php` and `src/engine/BMSkillTurbo.php`, how difficult would that be?  Again, in some sense the log message changes are what's really grungy to regression-check, so if only logs are changing, i can crunch through that, and then we can focus on whether the game logic changes are doing the right thing.

**Comment by blackshadowshade** [comment]
It's not difficult for me to split this pull request up again, as long as I know exactly what I need in each portion.

**Comment by cgolubi1** [comment]
Okay, great.  Here's how i think we should split it:

First PR, all the logging text changes, no changes to game behavior, includes the changes to these files:
* src/engine/BMDie.php - this changes a function which should only affect logging, to use the new flag
* src/engine/BMGameAction.php - this is the logging changes themselves

Second PR, the game logic changes, no changes to how logs are displayed when the same thing happens:
* src/engine/BMSkillDoppelganger.php - this causes Doppelganger to mad/mood afterwards, a game logic change
* src/engine/BMSkillTurbo.php - this causes Turbo to mad/mood afterwards, a game logic change

And then put the test changes in with whichever PR they need to for the tests to pass --- my guess is that most of them will go with the first PR, and ideally there will be one or two that go with the second, or else we're under-testing around those game logic changes. 

No rush, and thank you!

**Comment by blackshadowshade** [comment]
I've created #2635 which has all the mood logging changes.

Changes to the doppelganger / turbo behaviour are to follow in a second pull request. When we're up to that, we'll probably need a responder test or two to lock in that behaviour.

**Comment by cgolubi1** [comment]
Now that i've merged #2635 go ahead and put up a PR with the Doppelganger / Turbo changes, and i'll take a look and see what we need.

Apologies for not having hit on the idea of asking you to split the PR years ago; it makes a huge difference in testing.

**Comment by blackshadowshade** [comment]
Excellent! Moving on! New pull request coming up soon.

**Comment by blackshadowshade** [comment]
Closing this because it's now superfluous.

**Pull request closed without merging**